### ISSUE TITLE
`use_longlora: true` leaks future tokens into earlier positions and applies no S² grouping: the Q/K roll wraps the last tokens into positions 0-1

### ISSUE THREAD

**Issue by MakazhanAlpamys** [MakazhanAlpamys/Soup#1240]
## What happens

`training.use_longlora: true` installs `apply_longlora_forward_override(model)` for the duration of SFT training. The override rolls the q/k projection outputs of half the heads along the sequence with `torch.roll`, which wraps the last `group_size // 2` positions around to the front. Attention stays full-causal. Every query can therefore see keys computed from the last tokens of the sequence.

CPU only, a tiny random-init Llama built from a config (no download), default `group_size=4`. Changing only the last token should affect only the last position's logits:

```python
import torch
from transformers import LlamaConfig, LlamaForCausalLM
from soup_cli.utils.longlora import apply_longlora_forward_override

T = 8
torch.manual_seed(0)
a = torch.randint(4, 64, (1, T))
b = a.clone()
b[0, -1] = (a[0, -1] + 1 - 4) % 60 + 4      # change ONLY the last token

for impl in ("eager", "sdpa"):
    cfg = LlamaConfig(vocab_size=64, hidden_size=32, intermediate_size=64,
                      num_hidden_layers=1, num_attention_heads=4, num_key_value_heads=4,
                      max_position_embeddings=64)
    cfg._attn_implementation = impl
    torch.manual_seed(0)
    m = LlamaForCausalLM(cfg).eval()
    with torch.no_grad():
        plain_a, plain_b = m(a).logits, m(b).logits
        with apply_longlora_forward_override(m):
            s2_a, s2_b = m(a).logits, m(b).logits
            att = m(a, output_attentions=True).attentions[0] if impl == "eager" else None

    def moved(x, y):
        return [t for t in range(T) if not torch.allclose(x[0, t], y[0, t], atol=1e-6)]

    print(f"[{impl}] plain attention : {moved(plain_a, plain_b)}")
    print(f"[{impl}] S2 override     : {moved(s2_a, s2_b)}")
    if att is not None:
        grp = torch.arange(T) // 4                       # group_size=4 -> [0..3], [4..7]
        cross = grp[:, None] != grp[None, :]
        causal = torch.ones(T, T).tril().bool()
        print("attention mass on keys in an EARLIER group, per head:",
              [round(float(v), 3) for v in att[0][:, cross & causal].sum(dim=-1)])
```

```
[eager] plain attention : [7]
[eager] S2 override     : [1, 2, 3, 4, 5, 6, 7]
attention mass on keys in an EARLIER group, per head: [2.542, 2.536, 2.538, 2.536]
[sdpa] plain attention : [7]
[sdpa] S2 override     : [1, 2, 3, 4, 5, 6, 7]
```

(Also measured: the largest logit difference between the override and plain attention on the same input was 0.0010 on this tiny model, and after the context exits the logits are bit-identical to plain attention, so the projections are restored correctly.)

Positions 1-6 change when only token 7 changes, so the training loss at those positions can read the sequence's final tokens. The attention-mass line sums, over the four queries of the second group, the weight on keys of the first group: 2.54 of a possible 4.0 on every head, including the two unshifted ones. Under S² it would be 0 for those heads. No grouped (local) attention happens at all, so S² also delivers none of its memory or compute saving.

## Why

`src/soup_cli/utils/longlora.py:301-306` (`shift_heads_for_s2`) rolls the second half of the heads by `group_size // 2` with `torch.roll`, which wraps around, and `longlora.py:503-506` applies it to the `q_proj` and `k_proj` outputs. That is before RoPE, and `v_proj` is left unshifted. `group_size` is used only to size that shift; no code anywhere splits attention into groups. `trainer/sft.py:2216-2221` installs the override around `trainer.train()`, and the adapter is saved at `sft.py:2243` after the context has restored plain attention. LongLoRA does train with S² and infer with full attention by design. Here, though, the training-time attention is not S²: it is full causal attention over misaligned Q/K/V with wrapped tail tokens.

## Fix path

**Implement S²** as in the paper (section 3.2) and its reference code. Apply RoPE first. Shift q, k and v of half the heads by `-group_size // 2`. Split the sequence into groups and attend causally within each group. Roll the output of the shifted heads back. That requires owning the attention computation, for example a registered attention function in transformers 5, rather than wrapping the projection modules.

**Or refuse** `use_longlora: true` at config parse until that exists, with a message that says so and points to `rope_scaling_type` with plain LoRA. A partial fix, such as zero-padding instead of wrapping, removes the leak but still is not S².

## Acceptance

- [ ] A CPU test: with the override active, changing only the last token leaves the logits at every earlier position unchanged. Currently positions 1-6 change.
- [ ] A CPU test with eager attention and `output_attentions=True`: for the unshifted heads, the attention mass on keys outside the query's group is exactly 0. For the shifted heads it is confined to the shifted groups.
- [ ] Variants: `eager` and `sdpa`; GQA (`num_key_value_heads < num_attention_heads`); Mistral, Qwen2 and Phi-3 (fused `qkv_proj`); a sequence length that is not a multiple of `group_size`; right- and left-padded batches with batch > 1.
- [ ] If the refuse option is chosen instead: a parse-time error naming `use_longlora`.
